src/modules/ekf2/EKF/aid_sources/range_finder/range_finder_consistency_check.cpp

```diff
@@ -36,56 +36,137 @@
+	_time_last_update_us = time_us;
+	_x(RangeFilter::z.idx) -= dt * vz;
+	_P(RangeFilter::z.idx, RangeFilter::z.idx) += dt * dt * vz_var;
+	_P(RangeFilter::terrain.idx, RangeFilter::terrain.idx) += _terrain_process_noise;
+
+	// iterate through both measurements (z and dist_bottom)
+	const Vector2f measurements{z, dist_bottom};
 
-	// Variance of the time derivative of a random variable: var(dz/dt) = 2*var(z) / dt^2
-	const float var = 2.f * dist_bottom_var / (dt * dt);
-	_innov_var = var + vz_var;
+	for (int measurement_idx = 0; measurement_idx < 2; measurement_idx++) {
 
-	const float normalized_innov_sq = (_innov * _innov) / _innov_var;
-	_test_ratio = normalized_innov_sq / (_gate * _gate);
-	_signed_test_ratio_lpf.setParameters(dt, _signed_test_ratio_tau);
-	const float signed_test_ratio = matrix::sign(_innov) * _test_ratio;
-	_signed_test_ratio_lpf.update(signed_test_ratio);
+		float R;
+		bool reject = false;
+		Vector2f H;
 
-	updateConsistency(vz, time_us);
+		if (measurement_idx == 0) {
+			// direct state measurement
+			H(RangeFilter::z.idx) = 1.f;
+			H(RangeFilter::terrain.idx) = 0.f;
+			R = z_var;
 
-	_time_last_update_us = time_us;
-	_dist_bottom_prev = dist_bottom;
+		} else if (measurement_idx == 1) {
+			H(RangeFilter::z.idx) = _Ht(0);
+			H(RangeFilter::terrain.idx) = _Ht(1);
+			R = dist_bottom_var;
+		}
+
+		// residual
+		const float measurement_pred = H * _x;
+		const float y = measurements(measurement_idx) - measurement_pred;
+
+		// for H as col-vector:
+		// innovation variance S = H^T * P * H + R
+		// kalman gain K = P * H / S
+		const float S = (H.transpose() * _P * H + R)(0, 0);
+		Vector2f K = (_P * H / S);
+
+		if (measurement_idx == 0) {
+			K(RangeFilter::z.idx) = 1.f;
+
+		} else if (measurement_idx == 1) {
+			_innov = y;
+			const float test_ratio = fminf(sq(y) / (sq(_gate) * S), 4.f); // limit to 4 to limit sensitivity to outliers
+			_test_ratio_lpf.update(sign(_innov) * test_ratio, dt);
+			reject = test_ratio > 1.f;
+		}
+
+		if (!reject) {
+			// update step
+			_x(RangeFilter::z.idx)       += K(RangeFilter::z.idx) * y;
+			_x(RangeFilter::terrain.idx) += K(RangeFilter::terrain.idx) * y;
+
+			// covariance update with Joseph form:
+			// P = (I - K H) P (I - K H)^T + K R K^T
+			Matrix2f I;
+			I.setIdentity();
+			Matrix2f IKH = I - K.multiplyByTranspose(H);
+			_P = IKH * _P * IKH.transpose() + (K * R).multiplyByTranspose(K);
+		}
+	}
+
+	evaluateState(dt, vz, vz_var);
 }
 
-void RangeFinderConsistencyCheck::updateConsistency(float vz, uint64_t time_us)
+void RangeFinderConsistencyCheck::evaluateState(const float dt, const float vz, const float vz_var)
 {
-	if (fabsf(_signed_test_ratio_lpf.getState()) >= 1.f) {
-		if ((time_us - _time_last_horizontal_motion) > _signed_test_ratio_tau) {
-			_is_kinematically_consistent = false;
-			_time_last_inconsistent_us = time_us;
+	// start the consistency check after 1s
+	if (_t_since_first_sample + dt > 1.0f) {
+		_t_since_first_sample = 2.0f;
```

Comment:
Do you really need to force it to 2.0? Maybe remove the +dt in the condition so when it enters the `if` once, it will always enter it (even if dt is suddenly slightly lower).